feat: Allow owner to disable guest links

x1uc opened the pull request:
**feat:** Allow owner to disable guest links

This PR implements the feature request in [#543](https://github.com/mtlynch/picoshare/issues/543).

To summarize what has been done:

- Added a column (`disabled`) to the `guest_links` table & the `GuestLink` struct.
- Updated the `isActive` check for guest links to include the `disabled` field.
- Added a column on the guest links front page, with a button to enable/disable the link.
- Added new APIs related to managing guest links' disabled state.

I’m a beginner, and I believe there may be many areas where my work lacks standardization from your perspective. Please point them out, and I will work on improving. I’m very glad to hear your suggestions, which will not only improve the quality of this project but also enhance my programming skills.

----
Screen
![output](https://github.com/user-attachments/assets/92278f4e-f4a6-4105-9058-c4b4b458dfbe)

mtlynch commented on store/sqlite/guest_links.go:
```diff
@@ -127,17 +129,62 @@ func (s Store) DeleteGuestLink(id picoshare.GuestLinkID) error {
+func (s Store) DisableGuestLink(id picoshare.GuestLinkID) error {
+	log.Printf("Disable guest link %s", id)
+
+	tx, err := s.ctx.BeginTx(context.Background(), nil)
```
Transactions are useful when we have multiple SQL statements, and we want to rollback previous statements if any in the transaction fail.

When there's only one statement, we can just execute it immediately, like in `InsertGuestLink`.

(ditto below)

mtlynch commented on store/sqlite/guest_links.go:
```diff
@@ -127,17 +129,62 @@ func (s Store) DeleteGuestLink(id picoshare.GuestLinkID) error {
 }
 
+func (s Store) DisableGuestLink(id picoshare.GuestLinkID) error {
+	log.Printf("Disable guest link %s", id)
```
Log messages are not capitalized. Also, for concistency, it should be "disabling" so this should look like:

```go
log.Printf("disabling guest link %s", id)
```

(ditto for the enable function)

mtlynch commented on handlers/guest_links.go:
```diff
@@ -63,6 +63,40 @@ func (s Server) guestLinksDelete() http.HandlerFunc {
 	}
 }
 
+func (s *Server) guestLinksDisable() http.HandlerFunc {
```
Can we add unit tests for this? See `TestDeleteExistingGuestLink` as an example.

(ditto for enable)

mtlynch commented on handlers/templates/pages/guest-link-index.html:
```diff
@@ -125,6 +173,7 @@ <h1 class="title">Guest Links</h1>
           <th>Max Upload Size</th>
           <th>Uploads</th>
           <th class="has-text-right">Actions</th>
+          <th>Enable/Disable</th>
```
Can disable/enable be one of the actions rather than having its own column? I think the ban icon would work:

https://fontawesome.com/icons/ban?s=solid

x1uc replied:
I think I need an icon to represent 'revoke ban'. Which icon would be a good opposite of the ban icon? I’ve looked at a few, but none of them seem very appropriate.

mtlynch replied:
I think the checkbox you're using now works well enough.

mtlynch commented on handlers/views.go:
```diff
@@ -90,6 +90,9 @@ func (s Server) guestLinkIndexGet() http.HandlerFunc {
 		"isActive": func(gl picoshare.GuestLink) bool {
 			return gl.IsActive()
 		},
+		"isDisabled": func(gl picoshare.GuestLink) bool {
```
We don't need to do this, as the template can just call the struct's .IsDisabled function directly.

`isActive` set the wrong precedent, so I deleted that in #647.

mtlynch commented on picoshare/guest_link.go:
```diff
@@ -9,22 +9,24 @@ type (
 	GuestUploadCountLimit   *int
-
-	GuestLink struct {
+	GuestUploadDisabled     *int
```
Custom types are useful when we want to restrict what value can be in the native type. For example, a `GuestLinkID` is different than a `string` because not all valid `string`s are valid `GuestLinkID`s, so we have a type to restrict the values.

With a boolean, there's not much value in a custom type, so I think we should just add a boolean type to the `GuestLink` struct.

SQLite has to store it as an integer because it doesn't have a boolean type, but everything outside of the sqlite package should treat the value as a `bool`, and our `store/sqlite` package should be responsible for serializing/deserializing it to an integer for storage.

mtlynch commented on picoshare/guest_link.go:
```diff
@@ -9,22 +9,24 @@ type (
 		FileLifetime   FileLifetime
 		MaxFileBytes   GuestUploadMaxFileBytes
 		MaxFileUploads GuestUploadCountLimit
+		Disabled       GuestUploadDisabled
```
Can we make this `IsDisabled`? Our current convention is to use the `Is` prefix for `bool` variables.

mtlynch commented on store/sqlite/migrations/010-add-guest_links-enable-status.sql:
```diff
@@ -0,0 +1 @@
+ALTER TABLE guest_links ADD COLUMN disabled INTEGER;
```
Can we make this `is_disabled`, too?

mtlynch commented on handlers/templates/pages/guest-link-index.html:
```diff
@@ -82,6 +86,50 @@
+                  '<i class="fas fa-check-square" aria-hidden="true"></i>';
+                document
+                  .querySelector("snackbar-notifications")
+                  .addInfoMessage("Successfully disabled the guest link.");
```
The snackbar messages are flipped, so it should say enabled here and disabled below.

mtlynch commented on handlers/templates/pages/guest-link-index.html:
```diff
@@ -29,7 +29,11 @@
 
 {{ define "script-tags" }}
```
Can we add an e2e test to cover this new functionality to `e2e/guest-links.spec.ts`?

mtlynch commented on handlers/routes.go:
```diff
@@ -14,6 +14,8 @@ func (s *Server) routes() {
 	authenticatedApis.HandleFunc("/guest-links", s.guestLinksPost()).Methods(http.MethodPost)
 	authenticatedApis.HandleFunc("/guest-links/{id}", s.guestLinksDelete()).Methods(http.MethodDelete)
+	authenticatedApis.HandleFunc("/guest-links/enable/{id}", s.guestLinksEnable()).Methods(http.MethodPatch)
+	authenticatedApis.HandleFunc("/guest-links/disable/{id}", s.guestLinksDisable()).Methods(http.MethodPatch)
```
I think `/guest-links/{id}/enable` is a bit more idiomatic and consistent with our other routes.

Also, these seem like they should be `PUT` rather than `PATCH` as they're idempotent:

>In comparison with [PUT](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/PUT), a PATCH serves as a set of instructions for modifying a resource, whereas PUT represents a complete replacement of the resource. A PUT request is always [idempotent](https://developer.mozilla.org/en-US/docs/Glossary/Idempotent) (repeating the same request multiple times results in the resource remaining in the same state), whereas a PATCH request may not always be idempotent. For instance, if a resource includes an auto-incrementing counter, a PUT request will overwrite the counter (since it replaces the entire resource), but a PATCH request may not.

https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/PATCH

mtlynch commented on handlers/guest_links.go:
```diff
@@ -63,6 +63,40 @@ func (s Server) guestLinksDelete() http.HandlerFunc {
+			log.Printf("failed to delete guest link: %v", err)
+			http.Error(w, fmt.Sprintf("Failed to delete guest link: %v", err), http.StatusInternalServerError)
+			return
+		}
```
I haven't been good about doing this elsewhere, but if a handler never returns any content on success, it should end with:

```go
w.WriteHeader(http.StatusNoContent)
```

And we should check for the correct status code in unit tests.

mtlynch commented on handlers/templates/pages/guest-link-index.html:
```diff
@@ -82,6 +86,50 @@
+          if (toggleBtn.getAttribute("aria-label") === "Enable") {
+            guestLinkEnable(id)
+              .then(() => {
+                const button = toggleBtn;
```
Why are we defining a new variable? Couldn't we just keep using `toggleBtn`?

mtlynch commented on handlers/templates/pages/guest-link-index.html:
```diff
@@ -82,6 +86,50 @@
+                button.classList.remove("is-danger");
+                button.classList.add("is-success");
+                button.setAttribute("aria-label", "Disable");
+                button.innerHTML =
```
I try to avoid assigning `.innerHTML` since it's typically more powerful than we need. In this case, it looks like the only change is adding or removing `fa-check-square`. Can we do that with `classList` instead?

mtlynch commented on handlers/templates/pages/guest-link-index.html:
```diff
@@ -82,6 +86,50 @@
+        toggleBtn.addEventListener("click", () => {
+          const id = toggleBtn.getAttribute("pico-link-id");
+          // switch enable/disable
+          if (toggleBtn.getAttribute("aria-label") === "Enable") {
```
The `if-else` in the loop body means that the Enable and Disable cases share basically no code. I think we should either:

1. Use [classList.toggle](https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList/toggle) to share code outside the if-else
2. Split it up into two separate loops with no `if-else`.

mtlynch reviewed (changes requested):
Great work! You did a good job of matching a lot of the conventions in the codebase so that your changes fit in elegantly.

I added some notes on places where I think we could improve, but I think this is off to a great start.

x1uc commented:
I apologize for the two unsuccessful pushes on CircleCI. Will Circle send you any email notifications about the CI failure and disturb you?
I noticed that the CONTRIBUTE.md mentions using `npm run format` to format the frontend code, but it doesn't seem to work. I think the script should be added to the `package.json` like this:
```json
"scripts": {
  "format": "prettier --write ."
}
```

x1uc commented on picoshare/guest_link.go:
```diff
@@ -55,12 +54,12 @@ func (gl GuestLink) IsExpired() bool {
 
-func (gl GuestLink) IsDisabled() bool {
-	return gl.Disabled != GuestLinkDisabled && *gl.Disabled == 1
+func (gl GuestLink) IsDisabledLink() bool {
```
It seems that this function isn't actually needed, I only realized that after reviewing it later. I read your article on code reviews, and you mentioned that submitting code at night can sometimes cause trouble, and I think that's true.

mtlynch commented:
>I apologize for the two unsuccessful pushes on CircleCI. Will Circle send you any email notifications about the CI failure and disturb you?

No, it's fine. I don't receive any of those emails, so feel free to build as much as you'd like.

>I noticed that the CONTRIBUTE.md mentions using npm run format to format the frontend code, but it doesn't seem to work.

Ah, yes, thanks for letting me know. I've fixed the instructions in #648.

x1uc commented:
**Changelist:**  
- Rename the field from `diabled` to `is_disabled` .  
- Replace `innerHtml` manipulation with `classList.toggle` for cleaner DOM updates.  
- Refactor the API to follow RESTful conventions.  
- Implement end-to-end tests for this feature.  
- Add unit tests for `guestLinksDisable` and `guestLinksEnable` functions.

mtlynch commented:
Thanks for updating! I started a review, but Github seems to be having an outage that's preventing me from adding more notes to the review. I'll try again tomorrow.

mtlynch commented on handlers/templates/pages/guest-link-index.html:
```diff
@@ -82,6 +86,60 @@
+      },
+      Disable: {
+        apiCall: guestLinkDisable,
+        message: "Successfully disabled the guest link.",
```
Can we make this "Guest link disabled" to match the structure of the "enable" message?

mtlynch commented on handlers/templates/pages/guest-link-index.html:
```diff
@@ -82,6 +86,60 @@
+    const guestLinkButtonStates = {
+      Enable: {
+        apiCall: guestLinkEnable,
+        message: "Guest link enabled successfully.",
```
I think we can cut out "successfully" for brevity.

mtlynch reviewed (changes requested):
Great work on adding tests! You're continuing to match the conventions of the codebase really well and fit the feature to PicoShare's existing architecture.

I have a few smaller notes, but we're getting close to the finish line.

x1uc commented:
**Change List**  
- Refactored tests to use the `if got, want` style.
- Refactored unit tests to follow the `table-driven test` approach.  
- Made several grammatical corrections.
- Updated response status for invalid or non-existent `guest_link id` from `204 No Content` to `404 Not Found` .

mtlynch reviewed (changes requested):
Nice work! Just a little tidying up in the unit test, and then we should be ready to merge.

x1uc commented on handlers/guest_links_test.go:
```diff
@@ -346,3 +345,253 @@ func TestDeleteInvalidGuestLink(t *testing.T) {
+
+func TestGuestLinkToggleInValidRequest(t *testing.T) {
+
+	type operation struct {
```
Yes, I read the article, and it makes sense. The issue in `TestGuestLinkOperations` can be easily avoided by modifying the code as follows:
```go
[]struct {
    description string
    url         string
    expected    int
}{
    {
        description: "disable a nonexistent guest link",
        url:         "/api/guest-links/abcdefgh23456789/disable",
        expected:    http.StatusNotFound,
    },
}
```

However, the function `TestGuestLinkToggleInvalidRequest` also uses `fmt.Sprintf("/api/guest-links/%s/%s", id, action)` to construct URLs dynamically because the `guest_link id` is uncertain before the test runs. I'm not sure how to address this. Could you provide some advice on how to improve or refactor the function?

Additionally, I’m new to testing. In the codebase, I couldn’t find any indirect functions (like createLink or createOperation) being used in tests. Is it acceptable to use these indirect functions in tests, or is it better to avoid them for better test practices?

x1uc commented:
### Change List:
- Removed the `createLink` and `createOperation` methods.
- Eliminate uncertain test values.
- Replaced concatenated strings with direct URLs.
- Changed HTTP response status for invalid `guest_link id` from `404 Not Found` to `400 Bad Request`.

mtlynch commented on handlers/guest_links_test.go:
```diff
@@ -346,3 +345,193 @@ func TestDeleteInvalidGuestLink(t *testing.T) {
+				UrlExpires:     mustParseExpirationTime("2030-01-02T03:04:25Z"),
+				FileLifetime:   picoshare.FileLifetimeInfinite,
+				MaxFileBytes:   picoshare.GuestUploadUnlimitedFileSize,
+				MaxFileUploads: picoshare.GuestUploadUnlimitedFileUploads,
```
Can we include the `IsDisabled` field to make it clear what the starting state is?

mtlynch commented on handlers/guest_links_test.go:
```diff
@@ -346,3 +345,193 @@ func TestDeleteInvalidGuestLink(t *testing.T) {
 	}
 }
+
+func TestGuestLinkToggleInValidRequest(t *testing.T) {
```
The names seem to be flipped because these other test is called "TestGuestLinkOperations" but that's testing invalid input.

mtlynch commented on handlers/guest_links_test.go:
```diff
@@ -346,3 +345,193 @@ func TestDeleteInvalidGuestLink(t *testing.T) {
+	}
+}
+
+func TestGuestLinkOperations(t *testing.T) {
```
Is it possible to merge this into the test above? I just did this with the creation tests:

https://github.com/mtlynch/picoshare/blob/37fb930a8b99dbe49f20b8cdd4dfd27a467a5981/handlers/guest_links_test.go#L20-L275

x1uc replied:
Do you mean merging this test into TestGuestLinksPost? Hmm, I’m not sure about that, as I believe TestGuestLinksPost is specifically for testing POST /api/guest-links. Could you offer some suggestions on how to approach this?

mtlynch replied:
Sorry, no I'm saying merge together `TestGuestLinkToggleInValidRequest` and `TestGuestLinkOperations`

mtlynch commented on handlers/guest_links_test.go:
```diff
@@ -346,3 +345,193 @@ func TestDeleteInvalidGuestLink(t *testing.T) {
+
+	for _, tt := range []testCase{
+		{
+			description: "guest_link operation: disable",
```
Can we make the descriptions be what we are saying the endpoint does. For example, for this one, I'd say, "disabling active guest link succeeds"

mtlynch commented on handlers/guest_links_test.go:
```diff
@@ -346,3 +345,193 @@ func TestDeleteInvalidGuestLink(t *testing.T) {
+			},
+		},
+		{
+			description: "guest_link operation: enable",
```
I'd describe this one as "enabling guest link succeeds but has no effect when guest link is not disabled"

mtlynch reviewed (changes requested):
Just a few small things on the test to wrap up.

I know this has been a long review, so if you'd prefer me to do the last bit of tidying, I can do that.

x1uc commented:
**ChangeList:**  
0. Define struct like otherwise.  
1. Add `is_disabled` field in test struct.  
2. Do a single API request in each test case.  
3. Refactor: unify two tests into a single one.  
4. Change test description.

mtlynch commented on handlers/guest_links_test.go:
```diff
@@ -347,3 +346,164 @@ func TestDeleteInvalidGuestLink(t *testing.T) {
 }
+
+func TestEnableDisableGuestLink(t *testing.T) {
+
```
(nit) Can we delete this blank line?

mtlynch commented on handlers/guest_links_test.go:
```diff
@@ -347,3 +346,164 @@ func TestDeleteInvalidGuestLink(t *testing.T) {
+		t.Run(tt.description, func(t *testing.T) {
+			dataStore := test_sqlite.New()
+
+			if tt.status == http.StatusNoContent {
```
Can we change this to `if !tt.guestLinkInStore.Empty() {`?

I had trouble understand at first why we'd condition based on the HTTP response code when we hadn't even sent the HTTP request. I think checking based on whether the guest link is populated is the more direct way of testing whether we should add it into the mock store.

mtlynch commented on handlers/guest_links_test.go:
```diff
@@ -347,3 +346,164 @@ func TestDeleteInvalidGuestLink(t *testing.T) {
+			res := rec.Result()
+
+			if got, want := res.StatusCode, tt.status; got != want {
+				t.Errorf("%s: handler returned wrong status code: got %v want %v",
```
This should be `t.Fatalf`. If the status code is wrong, there's no sense in checking the response data.

And can we simplify to this to match the other tests?

```go
if got, want := res.StatusCode, tt.status; got != want {
  t.Fatalf("status=%d, want=%d", got, want)
}
```

Including the description is an artifact from before I realized we could do `t.Run`.

mtlynch commented on handlers/guest_links_test.go:
```diff
@@ -347,3 +346,164 @@ func TestDeleteInvalidGuestLink(t *testing.T) {
+			}
+
+			if got, want := gl, tt.expected; !reflect.DeepEqual(got, want) {
+				t.Fatalf("guestLink=%+v, want=%+v", got, want)
```
This should just be `t.Errorf` as it's not fatal for the result to not match expectations.

It's kind of moot since it's the last assertion in the test, but just good for consistency.

x1uc replied:
Makes sense. This might involve another error. I wrote this code to mimic this one:
https://github.com/mtlynch/picoshare/blob/37fb930a8b99dbe49f20b8cdd4dfd27a467a5981/handlers/guest_links_test.go#L270-L272

mtlynch commented on handlers/templates/pages/guest-link-index.html:
```diff
@@ -82,6 +86,58 @@
+      const oldIcon = button.querySelector("i");
+      const iconClasses = oldIcon.classList;
+      iconClasses.toggle("fa-check-square");
+      iconClasses.toggle("fa-ban");
```
Can we make this

```javascript
button.querySelector("i").toggle("fa-check-square");
button.querySelector("i").toggle("fa-ban");
```

I think I gave this note in a previous round but it might have been dropped.

x1uc replied:
Yes, I’m very sorry. You previously pointed out another issue, and I didn’t notice similar errors elsewhere. My apologies.

mtlynch reviewed (changes requested):
Looks good! Just a a small bit of tidying left.

mtlynch reviewed (approved):
Thanks so much! This looks good. I'm just going to go over it with some small tweaks, but I should be able to merge this in in the next few days.
